### PR TITLE
Upgrade gorm to v1.21.9

### PR DESCRIPTION
**Pull request by milton0825**
# TL;DR
Upgrade gorm to v1.21.9 so that we can use new features such as upsert

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [ ] Smoke tested
 - [X] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

A few backward incompatible changes from gorm upgrade:
- `association_foreignkey` renamed to `references`. [v1](https://v1.gorm.io/docs/belongs_to.html#Association-ForeignKey), [v2](https://gorm.io/docs/belongs_to.html#Override-References)
- `offset` and `limit` changed from `uint32` to `int`
- DB connection API interface changed
- Import path change: `github.com/jinzhu/gorm` to `gorm.io/gorm`

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1027

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

**Comment by milton0825** [comment]
@kumare3  can you approve workflow?

File: pkg/repositories/gormimpl/artifact.go [hunk 1]
```diff
@@ -65,18 +65,20 @@ func (h *artifactRepo) Get(ctx context.Context, in models.ArtifactKey) (models.A
...
-		return models.Artifact{}, errors.GetMissingEntityError("Artifact", &datacatalog.Artifact{
-			Dataset: &datacatalog.DatasetID{
-				Project: in.DatasetProject,
-				Domain:  in.DatasetDomain,
-				Name:    in.DatasetName,
-				Version: in.DatasetVersion,
-			},
-			Id: in.ArtifactID,
-		})
+		switch result.Error.Error() {
```
Comment (katrogan) [inline pkg/repositories/gormimpl/artifact.go:68]: do you need a switch here? why not just an if statement?

File: pkg/repositories/gormimpl/artifact_test.go [hunk 2]
```diff
@@ -117,21 +117,22 @@ func TestCreateArtifact(t *testing.T) {
...
+		`INSERT INTO "artifacts" ("created_at","updated_at","deleted_at","dataset_project","dataset_name","dataset_domain","dataset_version","artifact_id","dataset_uuid","serialized_metadata") VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)`).WithCallback(
 		func(s string, values []driver.NamedValue) {
 			artifactCreated = true
 		},
 	)
 
 	GlobalMock.NewMock().WithQuery(
-		`INSERT  INTO "artifact_data" ("created_at","updated_at","deleted_at","dataset_project","dataset_name","dataset_domain","dataset_version","artifact_id","name","location") VALUES (?,?,?,?,?,?,?,?,?,?)`).WithCallback(
+		`INSERT INTO "artifact_data" ("created_at","updated_at","deleted_at","dataset_project","dataset_name","dataset_domain","dataset_version","artifact_id","name","location") VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10),($11,$12,$13,$14,$15,$16,$17,$18,$19,$20) ON CONFLICT ("dataset_project","dataset_name","dataset_domain","dataset_version","artifact_id","name") DO UPDATE SET "dataset_project"="excluded"."dataset_project","dataset_name"="excluded"."dataset_name","dataset_domain"="excluded"."dataset_domain","dataset_version"="excluded"."dataset_version","artifact_id"="excluded"."artifact_id"`).WithCallback(
```
Comment (katrogan) [inline pkg/repositories/gormimpl/artifact_test.go:127]: interesting, so gorm behavior has joined to do an upsert for relations?

File: pkg/repositories/gormimpl/artifact_test.go [hunk 3]
```diff
@@ -175,13 +176,13 @@ func TestGetArtifact(t *testing.T) {
...
 	GlobalMock.NewMock().WithQuery(
-		`SELECT * FROM "artifact_data"  WHERE "artifact_data"."deleted_at" IS NULL AND ((("dataset_project","dataset_name","dataset_domain","dataset_version","artifact_id") IN ((testProject,testName,testDomain,testVersion,123)))) ORDER BY "artifact_data"."dataset_project" ASC`).WithReply(expectedArtifactDataResponse)
+		`SELECT * FROM "artifact_data" WHERE ("artifact_data"."dataset_project","artifact_data"."dataset_name","artifact_data"."dataset_domain","artifact_data"."dataset_version","artifact_data"."artifact_id") IN (($1,$2,$3,$4,$5))%!!(string=123)!(string=testVersion)!(string=testDomain)!(string=testName)(EXTRA string=testProject)`).WithReply(expectedArtifactDataResponse)
 	GlobalMock.NewMock().WithQuery(
-		`SELECT * FROM "partitions"  WHERE "partitions"."deleted_at" IS NULL AND (("artifact_id" IN (123))) ORDER BY partitions.created_at ASC,"partitions"."dataset_uuid" ASC`).WithReply(expectedPartitionResponse)
+		`SELECT * FROM "partitions" WHERE "partitions"."artifact_id" = $1 ORDER BY partitions.created_at ASC%!(EXTRA string=123)`).WithReply(expectedPartitionResponse)
```
Comment (katrogan) [inline pkg/repositories/gormimpl/artifact_test.go:183]: this EXTRA... seems like a golang formatting issue with a mismatch in placeholders and format args. are you sure this is producing the intended query?

File: pkg/repositories/gormimpl/artifact_test.go [hunk 4]
```diff
@@ -175,13 +176,13 @@ func TestGetArtifact(t *testing.T) {
 
 	// Only match on queries that append expected filters
 	GlobalMock.NewMock().WithQuery(
-		`SELECT * FROM "artifacts"  WHERE "artifacts"."deleted_at" IS NULL AND (("artifacts"."dataset_project" = testProject) AND ("artifacts"."dataset_name" = testName) AND ("artifacts"."dataset_domain" = testDomain) AND ("artifacts"."dataset_version" = testVersion) AND ("artifacts"."artifact_id" = 123)) ORDER BY artifacts.created_at DESC,"artifacts"."dataset_project" ASC LIMIT 1`).WithReply(expectedArtifactResponse)
+		`SELECT * FROM "artifacts" WHERE "artifacts"."dataset_project" = $1 AND "artifacts"."dataset_name" = $2 AND "artifacts"."dataset_domain" = $3 AND "artifacts"."dataset_version" = $4 AND "artifacts"."artifact_id" = $5 ORDER BY artifacts.created_at DESC,"artifacts"."created_at" LIMIT 1%!!(string=123)!(string=testVersion)!(string=testDomain)!(string=testName)(EXTRA string=testProject)`).WithReply(expectedArtifactResponse)
```
Comment (katrogan) [inline pkg/repositories/gormimpl/artifact_test.go:179]: what does the `%!!` mean?

**Review by katrogan — COMMENTED** [review]
Unfortunately I don't have permissions to enable CI workflow runs for this PR (cc @kumare3 or @EngHabu do you?) @milton0825 can you run integration tests locally for this PR and verify they succeed?

File: pkg/repositories/gormimpl/artifact.go (hunk 1 above)
Comment (milton0825) [reply pkg/repositories/gormimpl/artifact.go:68]: Both should work. If we want to handle a different error type, probably less code change needed.

File: pkg/repositories/gormimpl/artifact_test.go (hunk 2 above)
Comment (milton0825) [reply pkg/repositories/gormimpl/artifact_test.go:127]: Correct.

File: pkg/repositories/gormimpl/artifact_test.go (hunk 3 above)
Comment (milton0825) [reply pkg/repositories/gormimpl/artifact_test.go:183]: Yeah these are the query template and the arguments and that is what the mock framework captures.

File: pkg/repositories/gormimpl/artifact_test.go (hunk 4 above)
Comment (milton0825) [reply pkg/repositories/gormimpl/artifact_test.go:179]: I believe that is how it separates the query template and the arguments.

File: pkg/repositories/gormimpl/artifact_test.go (hunk 3 above)
Comment (milton0825) [reply pkg/repositories/gormimpl/artifact_test.go:183]: It should be rendering the query template somewhere deeper but the mocking framework we are using is not capturing that.

**Comment by milton0825** [comment]
@katrogan unit & lint tests passed locally.

**Comment by katrogan** [comment]
@milton0825 sorry I thought this repo had integration tests too, but I guess not. This is a little tedious, but would you mind doing the following to test:

1. for this repo, running `make docker_build` to build a docker image locally
2. Clone the flyte repo and run `make deploy_sandbox`
3. edit the datacatalog deployment `kubectl -n flyte edit deployment datacatalog` to change instances of `image: ghcr.io/flyteorg/datacatalog:v0.3.1` to `datacatalog:<your local image sha`
4. register the sample workflows in [flytesnacks](https://github.com/flyteorg/flytesnacks):  `make register`
5. Open the flyteconsole `localhost:30081/console`,  choose the `flytetester` project and `development` domain, and navigate to the Tasks sidebar in the left hand side and run this [square](https://github.com/flyteorg/flytesnacks/blob/master/cookbook/core/flyte_basics/task_cache.py#L29) task a few times with identical input and verify subsequent runs read from the cache

File: pkg/repositories/gormimpl/artifact.go (hunk 1 above)
Comment (katrogan) [reply pkg/repositories/gormimpl/artifact.go:68]: hm switch seems like overkill still and this code doesn't really change

File: pkg/repositories/gormimpl/artifact_test.go (hunk 3 above)
Comment (katrogan) [reply pkg/repositories/gormimpl/artifact_test.go:183]: ah it looks like go-mocket hasn't been released in 3 years, makes sense

**Comment by milton0825** [comment]
I think `make register` may not work with the sandbox created via step 2:
```
➜ make register
Makefile:43: *** Cluster has not been started! Use 'make start' to start a cluster.  Stop.
```
Maybe we need to use the sandbox created via https://github.com/flyteorg/flytesnacks/blob/master/Makefile#L68

Any idea about the difference between the sandbox environment in `flyte` vs `flytesnack` repo?

**Comment by milton0825** [comment]
Uh don't think we can easily use the sandbox environment created from https://github.com/flyteorg/flytesnacks/blob/master/Makefile#L47-L60 to run integration tests with datacatalog since it creates a docker container and does k8s deployment within container [here](https://github.com/flyteorg/flyte/blob/master/docker/sandbox/flyte-entrypoint-dind.sh#L34). 

Actually we may be able to sync the code to the container and build a docker image within the container..

```
docker exec -it flyte-sandbox bash
bash-5.1# docker images
REPOSITORY                        TAG                            IMAGE ID       CREATED        SIZE
ghcr.io/flyteorg/flytecookbook    core-latest                    a1e1c6f6b7b8   10 hours ago   1.26GB
ghcr.io/flyteorg/flyteadmin       v0.5.4                         a60afc88d0b8   9 days ago     62.4MB
ghcr.io/flyteorg/flytepropeller   v0.10.7                        bec6738fb13b   2 weeks ago    143MB
redocly/redoc                     latest                         8d9967d278da   2 weeks ago    33.8MB
postgres                          10.16-alpine                   e5ac6f3b3b4d   4 weeks ago    72.7MB
minio/minio                       RELEASE.2021-04-06T23-11-00Z   9ea160f24871   5 weeks ago    183MB
ghcr.io/flyteorg/datacatalog      v0.3.2                         2bc551042102   5 weeks ago    45.4MB
ghcr.io/flyteorg/flyteconsole     v0.19.7                        05e3c2418b7d   8 weeks ago    162MB
projectcontour/contour            v1.13.1                        794cdb4d725a   2 months ago   41.1MB
envoyproxy/envoy                  v1.17.1                        bdaff8e5dcc5   2 months ago   132MB
kubernetesui/dashboard            v2.2.0                         5c4ee6ca42ce   2 months ago   225MB
rancher/coredns-coredns           1.8.0                          296a6d5035e2   6 months ago   42.5MB
kubernetesui/metrics-scraper      v1.0.6                         48d79e554db6   6 months ago   34.5MB
rancher/pause                     3.1                            da86e6ba6ca1   3 years ago    742kB
```

**Comment by milton0825** [comment]
So the recommended way is to copy the repo to to the container and build the image within the container.
```
docker cp ~/go/src/github.com/flyteorg/datacatalog/ flyte-sandbox:/tmp/datacatalog
make shell
cd /tmp/datacatalog
make docker_build
```

**Comment by katrogan** [comment]
@milton0825 great! were you able to verify that caching works in your sandbox?

**Comment by milton0825** [comment]
Yeah I was able to run the integration tests and uncovered a few issues that I need to fix.

File: pkg/repositories/gormimpl/artifact.go (hunk 1 above)
Comment (milton0825) [reply pkg/repositories/gormimpl/artifact.go]: Fixed.

**Comment by milton0825** [comment]
I have tested the task caching in the sandbox environment:
First attempt, cache populated:
image
Second attempt, cache re-used:
image

File: pkg/repositories/gormimpl/artifact.go [hunk 5]
```diff
@@ -65,19 +66,20 @@ func (h *artifactRepo) Get(ctx context.Context, in models.ArtifactKey) (models.A
 		)
 
 	if result.Error != nil {
+		if result.Error.Error() == gorm.ErrRecordNotFound.Error() {
```
Comment (katrogan) [inline pkg/repositories/gormimpl/artifact.go:69]: I think this check is no longer necessary now that not found errors are returned in `result.Error` see https://github.com/flyteorg/datacatalog/blob/1a3b9195bb80ef23ea82eb25206e067afdf535af/pkg/repositories/errors/postgres.go#L30

here and in other places

File: pkg/repositories/gormimpl/list.go [hunk 6]
```diff
@@ -28,14 +37,21 @@ func applyListModelsInput(tx *gorm.DB, sourceEntity common.Entity, in models.Lis
 		return nil, errors.GetInvalidEntityError(sourceEntity)
 	}
 
-	sourceTableName := tx.NewScope(sourceModel).TableName()
+	sourceTableName, err := getTableName(tx, sourceModel)
+	if err != nil {
+		return nil, err
```
Comment (katrogan) [inline pkg/repositories/gormimpl/list.go:42]: can we wrap this in a DataCatalogError so we can add a grpc reponse code too? https://github.com/flyteorg/datacatalog/blob/master/pkg/errors/errors.go#L44:6

File: pkg/repositories/handle.go [hunk 7]
```diff
@@ -24,6 +24,9 @@ func NewDBHandle(dbConfigValues config.DbConfig, catalogScope promutils.Scope) (
 		User:         dbConfigValues.User,
 		Password:     dbConfigValues.Password,
 		ExtraOptions: dbConfigValues.ExtraOptions,
+		BaseConfig: config.BaseConfig{
+			DisableForeignKeyConstraintWhenMigrating: true,
```
Comment (katrogan) [inline pkg/repositories/handle.go:28]: why is this true?

**Review by katrogan — COMMENTED** [review]
sorry just a few more comments. thank you so much for testing this end to end!

File: pkg/repositories/gormimpl/artifact.go (hunk 5 above)
Comment (milton0825) [reply pkg/repositories/gormimpl/artifact.go:69]: I think the error message returned will be different. In the old GORM version, we are returning

File: pkg/repositories/handle.go (hunk 7 above)
Comment (milton0825) [reply pkg/repositories/handle.go:28]: So I believe that we do not populate foreign keys at all using GORM v1. Therefore I turn off populating foreign key during migration in GORM v2 to match the behavior of v1. I think we can address that in a separate PR.

Here are the table constraints populated by v1 gorm and you can see no foreign constraint is created.
```
datacatalog=> select table_name, constraint_name, constraint_type from information_schema.table_constraints where table_name in ('tags', 'artifacts', 'datasets');
 table_name |    constraint_name    | constraint_type
------------+-----------------------+-----------------
 datasets   | datasets_pkey         | PRIMARY KEY
 datasets   | datasets_uuid_key     | UNIQUE
 tags       | tags_pkey             | PRIMARY KEY
 artifacts  | artifacts_pkey        | PRIMARY KEY
 artifacts  | 2200_16421_4_not_null | CHECK
 artifacts  | 2200_16421_5_not_null | CHECK
 artifacts  | 2200_16421_6_not_null | CHECK
 artifacts  | 2200_16421_7_not_null | CHECK
 artifacts  | 2200_16421_8_not_null | CHECK
 datasets   | 2200_16412_4_not_null | CHECK
 datasets   | 2200_16412_5_not_null | CHECK
 datasets   | 2200_16412_6_not_null | CHECK
 datasets   | 2200_16412_7_not_null | CHECK
 tags       | 2200_16439_4_not_null | CHECK
 tags       | 2200_16439_5_not_null | CHECK
 tags       | 2200_16439_6_not_null | CHECK
 tags       | 2200_16439_7_not_null | CHECK
 tags       | 2200_16439_8_not_null | CHECK
 tags       | 2200_16439_9_not_null | CHECK
(19 rows)
```

File: pkg/repositories/gormimpl/list.go (hunk 6 above)
Comment (milton0825) [reply pkg/repositories/gormimpl/list.go:44]: Fixed.

File: pkg/repositories/gormimpl/artifact.go (hunk 5 above)
Comment (katrogan) [reply pkg/repositories/gormimpl/artifact.go:69]: true, the response code is the same. not really a fan of having to duplicate this code everywhere though seeing as we already have an error transformer. maybe we can pass in a formatted entity identifier to the error transformer so it can customize the error message

File: pkg/repositories/handle.go (hunk 7 above)
Comment (katrogan) [reply pkg/repositories/handle.go:28]: ah interesting, mind filing an issue in https://github.com/flyteorg/flyte to explore enabling it in a separate PR?

**Review by katrogan — APPROVED** [review]
thank you for this! mind waiting one last time for a maintainer (cc @EngHabu) to run the PR checks?

**Comment by katrogan** [comment]
hey @milton0825 looks like lint is failing, mind running `make goimports` and pushing?

**Comment by milton0825** [comment]
@katrogan fixed lint

**Comment by milton0825** [comment]
For the code coverage, it is mainly because `Migrate` func in `migrate.go` not covered. And unfortunately, the sql mock framework does not support mocking `CREATE` statement [1].

```
unimplemented statement Exec command type of "CREATE"
```

[1] https://github.com/Selvatico/go-mocket#go-mocket--go-gorm--sql-mocking-library

**Review by katrogan — APPROVED** [review]
